flink-kubernetes/src/main/java/org/apache/flink/kubernetes/highavailability/KubernetesStateHandleStore.java

```diff
@@ -496,12 +498,14 @@ public boolean releaseAndTryRemove(String key) throws Exception {
                                     Objects.requireNonNull(configMap.getData().remove(key));
                                 }
                                 return Optional.of(configMap);
+                            } else {
+                                stateHandleDoesNotExist.set(true);
                             }
                             return Optional.empty();
                         })
                 .thenCompose(
                         updated -> {
-                            if (updated && stateHandleRefer.get() != null) {
+                            if (stateHandleRefer.get() != null) {
```

Comment:
Why are we now assuming that update == true if the ref isn't null? Even if it's currently the case, that seems to rely on implementation details of the KubeClient.

Reply:
That's a valid point. Even if we set the `stateHandleReferer` it could mean that the update did not happen and in that case, we don't want to discard the data. I will revert it